src/validation.cpp

```diff
@@ -2346,13 +2346,40 @@ bool CChainState::FlushStateToDisk(
                 }
                 std::vector<const CBlockIndex*> vBlocks;
                 vBlocks.reserve(setDirtyBlockIndex.size());
+                std::set<CBlockIndex*> setTrimmableBlockIndex(setDirtyBlockIndex);
                 for (std::set<CBlockIndex*>::iterator it = setDirtyBlockIndex.begin(); it != setDirtyBlockIndex.end(); ) {
                     vBlocks.push_back(*it);
                     setDirtyBlockIndex.erase(it++);
                 }
                 if (!pblocktree->WriteBatchSync(vFiles, nLastBlockFile, vBlocks)) {
                     return AbortNode(state, "Failed to write to block index database");
                 }
+
+                if (fTrimHeaders) {
+                    LogPrintf("Flushing block index, trimming headers, setTrimmableBlockIndex.size(): %d\n", setTrimmableBlockIndex.size());
+                    int trim_height = m_chain.Height() - nMustKeepFullHeaders;
+                    int min_height = std::numeric_limits<int>::max();
+                    CBlockIndex* min_index = nullptr;
+                    for (std::set<CBlockIndex*>::iterator it = setTrimmableBlockIndex.begin(); it != setTrimmableBlockIndex.end(); it++) {
+                        (*it)->assert_untrimmed();
+                        if ((*it)->nHeight < trim_height) {
+                            (*it)->trim();
+                            if ((*it)->nHeight < min_height) {
+                                min_height = (*it)->nHeight;
+                                min_index = *it;
+                            }
+                        }
+                    }
+
+                    // Handle any remaining untrimmed blocks that were too recent for trimming last time we flushed.
+                    if (min_index) {
+                        min_index = min_index->pprev;
+                        while (min_index && !min_index->trimmed()) {
+                            min_index->trim();
+                            min_index = min_index->pprev;
+                        }
```

Comment:
Is this the logic which trims blocks that were new when we got them, but are now old enough that they should be trimmed?

It looks like it only gets triggered when we flush to disk. How often does this flushing take place?

Reply:
Correct. Adjusting the flush frequency is part of this PR elsewhere -- I think I lowered it from 1 hour to 5 minutes. It's not ideal for it to be time-based instead of memory-pressure-based, but that would be a more invasive change.